File: packages/patrol/android/src/main/kotlin/pl/leancode/patrol/PatrolAppServiceClient.java
package pl.leancode.patrol;

import io.grpc.Grpc;
import io.grpc.InsecureChannelCredentials;
import io.grpc.ManagedChannel;
import io.grpc.StatusRuntimeException;
import pl.leancode.patrol.contracts.PatrolAppServiceGrpc;

import static pl.leancode.patrol.contracts.Contracts.*;

/**
 * Enables querying Dart tests, running them, waiting for them to finish, and getting their results
 */
public class PatrolAppServiceClient {

    private final PatrolAppServiceGrpc.PatrolAppServiceBlockingStub blockingStub;

    public PatrolAppServiceClient() {
        String target = "localhost:8082"; // TODO: Document this value better
        Logger.INSTANCE.i("Created PatrolAppServiceClient with default target: " + target);
        ManagedChannel channel = Grpc.newChannelBuilder(target, InsecureChannelCredentials.create()).build();
        blockingStub = PatrolAppServiceGrpc.newBlockingStub(channel);
    }

    public PatrolAppServiceClient(String target) {
        Logger.INSTANCE.i("Created PatrolAppServiceClient, target: " + target);
        ManagedChannel channel = Grpc.newChannelBuilder(target, InsecureChannelCredentials.create()).build();
        blockingStub = PatrolAppServiceGrpc.newBlockingStub(channel);
    }

    public DartTestGroup listDartTests() throws StatusRuntimeException {
        Empty request = Empty.newBuilder().build();
        ListDartTestsResponse response = blockingStub.listDartTests(request);

        return response.getGroup();
    }

    public RunDartTestResponse runDartTest(String name) throws StatusRuntimeException {
        RunDartTestRequest request = RunDartTestRequest.newBuilder().setName(name).build();
        return blockingStub.runDartTest(request);
    }
}

File: packages/patrol/android/src/main/kotlin/pl/leancode/patrol/PatrolAppServiceClient.kt
package pl.leancode.patrol

import pl.leancode.patrol.contracts.Contracts
import pl.leancode.patrol.contracts.PatrolAppServiceClientException
import java.util.concurrent.TimeUnit
import pl.leancode.patrol.contracts.PatrolAppServiceClient as Client

/**
 * Enables querying Dart tests, running them, waiting for them to finish, and getting their results
 */
class PatrolAppServiceClient {

    private var client: Client

    // https://github.com/leancodepl/patrol/issues/1683
    private val timeout = 2L
    private val timeUnit = TimeUnit.HOURS

    constructor(port: Int) {
        client = Client(address = "localhost", port = port, timeout = timeout, timeUnit = timeUnit)
        Logger.i("Created PatrolAppServiceClient: ${client.serverUrl}")
    }

    constructor(address: String, port: Int) {
        client = Client(address = address, port = port, timeout = timeout, timeUnit = timeUnit)
        Logger.i("Created PatrolAppServiceClient: ${client.serverUrl}")
    }

    @Throws(PatrolAppServiceClientException::class)
    fun listDartTests(): Contracts.DartGroupEntry {
        Logger.i("PatrolAppServiceClient.listDartTests()")
        val result = client.listDartTests()
        return result.group
    }

    @Throws(PatrolAppServiceClientException::class)
    fun runDartTest(name: String): Contracts.RunDartTestResponse {
        Logger.i("PatrolAppServiceClient.runDartTest($name)")
        return client.runDartTest(Contracts.RunDartTestRequest(name))
    }
}
